cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+	// clusters to rollout
+	ClustersToRollout map[string]ClusterRolloutStatus
+	// clusters that are timeout
+	ClustersTimeOut map[string]ClusterRolloutStatus
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+}
+
+func NewRolloutHandler(pdTracker *clusterv1beta1.PlacementDecisionClustersTracker) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		pdTracker: pdTracker,
+	}
+}
+
+// The input is a duck type RolloutStrategy and a ClusterRolloutStatusFunc to return the rollout status on each managed cluster.
+// Return the strategy actual take effect and a list of cluster that needs to update with current state.
+func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	switch rolloutStrategy.Type {
+	case All:
+		return r.getRolloutAllClusters(rolloutStrategy, statusFunc)
+	case Progressive:
+		return r.getProgressiveClusters(rolloutStrategy, statusFunc)
+	case ProgressivePerGroup:
+		return r.getProgressivePerGroupClusters(rolloutStrategy, statusFunc)
+	default:
+		return nil, RolloutResult{}, fmt.Errorf("incorrect rollout strategy type %v", rolloutStrategy.Type)
+	}
+}
+
+func (r *RolloutHandler) getRolloutAllClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: All}
+	strategy.All = rolloutStrategy.All.DeepCopy()
+	if strategy.All == nil {
+		strategy.All = &RolloutAll{}
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.All.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// get all the clusters
+	totalClusters, _, totalClusterGroups := r.pdTracker.ExistingBesides([]clusterv1beta1.GroupKey{})
+	clusterToGroupKey := r.pdTracker.ClusterToGroupKey(totalClusterGroups)
+	rolloutresult := progressivePerCluster(totalClusters.UnsortedList(), clusterToGroupKey, len(totalClusters), failureTimeout, statusFunc)
+
+	return &strategy, rolloutresult, nil
+}
+
+func (r *RolloutHandler) getProgressiveClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: Progressive}
+	strategy.Progressive = rolloutStrategy.Progressive.DeepCopy()
+	if strategy.Progressive == nil {
+		strategy.Progressive = &RolloutProgressive{}
+	}
+
+	// upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
```

Comment:
So this section only returns the clusters in the mandatory decision groups? If so, I was not expecting this behavior from the `GetRolloutCluster` description, which calls this method.

Reply:
If there's mandatory decision groups defined and not all clusters upgrade successfully, `rolloutresult.ClustersToRollout` will contain the clusters that need to rollout and will return here. If not, it will continue to go through other groups.